src/Avalonia.Controls/Utils/SharedSizeScopeHost.cs

```diff
@@ -0,0 +1,400 @@
+                        NotifyNewItems();
+
+                        break;
+                }
+            }
+
+            public void UpdateMeasureResult(GridLayout.MeasureResult rowResult, GridLayout.MeasureResult columnResult)
+            {
+                MeasurementState = MeasurementState.Cached;
+                for (int i = 0; i < Grid.RowDefinitions.Count; i++)
+                {
+                    Results[i].MeasuredResult = rowResult.LengthList[i];
+                }
+
+                for (int i = 0; i < Grid.ColumnDefinitions.Count; i++)
+                {
+                    Results[i + rowResult.LengthList.Count].MeasuredResult = columnResult.LengthList[i];
+                }
+            }
+
+            public void InvalidateMeasure()
+            {
+                MeasurementState = MeasurementState.Invalidated;
+                Results.ForEach(r => r.MeasuredResult = double.NaN);
+            }
+
+            public void Dispose()
+            {
+                _subscriptions.Dispose();
+                _groupChanged.OnCompleted();
+            }
+
+            public Grid Grid { get; }
+            public MeasurementState MeasurementState { get; private set; }
+
+            public List<MeasurementResult> Results { get; private set; }
+        }
+
+        private readonly AvaloniaList<MeasurementCache> _measurementCaches;
+
+        private class MeasurementResult
+        {
+            public MeasurementResult(DefinitionBase definition)
+            {
+                Definition = definition;
+                MeasuredResult = double.NaN;
+            }
+
+            public DefinitionBase Definition { get; }
+            public double MeasuredResult { get; set; }
+        }
+
+        private class Group
+        {
+            public bool IsFixed { get; set; }
+
+            public List<MeasurementResult> Results { get; } = new List<MeasurementResult>();
+
+            public double CalculatedLength { get; }
+        }
+
+        private readonly Dictionary<string, Group> _groups = new Dictionary<string, Group>();
+
+
+        public SharedSizeScopeHost(Control scope)
+        {
+            _measurementCaches = GetParticipatingGrids(scope);
+
+            foreach (var cache in _measurementCaches)
+            {
+                cache.Grid.InvalidateMeasure();
+                AddGridToScopes(cache);
+
+                cache.GroupChanged.Subscribe(SharedGroupChanged);
+            }
+        }
+
+        void SharedGroupChanged((string oldName, string newName, MeasurementResult result) change)
+        {
+            RemoveFromGroup(change.oldName, change.result);
+            AddToGroup(change.newName, change.result);
+        }
+
+        internal void InvalidateMeasure(Grid grid)
+        {
+            var cache = _measurementCaches.FirstOrDefault(mc => ReferenceEquals(mc.Grid, grid));
+            Debug.Assert(cache != null);
+
+            cache.InvalidateMeasure();
+        }
```

Comment:
Should most likely also invalidate measure or arrange on the other participating grids.
My idea:
if, for a grid, _previousMeasure has any dimension infinite - InvalidateMeasure(), otherwise InvalidateArrange()